pkg/limits/ingest_limits.go

```diff
@@ -611,3 +624,34 @@ func (s *IngestLimits) GetStreamUsage(_ context.Context, req *logproto.GetStream
 		Rate:           uint64(rate),
 	}, nil
 }
+
+func (s *IngestLimits) streams(tenant string) <-chan *streamMetadata {
+	ch := make(chan *streamMetadata)
+
+	go func() {
+		for i := range s.metadata.size {
+			s.metadata.locks[i].RLock()
+
+			if s.metadata.stripes[i][tenant] == nil {
+				s.metadata.locks[i].RUnlock()
+				continue
+			}
+
+			for partitionID, streams := range s.metadata.stripes[i][tenant] {
+				if assigned := s.partitionManager.Has(partitionID); !assigned {
+					continue
+				}
+
+				for _, stream := range streams {
+					ch <- stream
+				}
+			}
+
+			s.metadata.locks[i].RUnlock()
+		}
+
+		close(ch)
+	}()
+
+	return ch
+}
```

Comment:
@grobinson-grafana Can it be that we need to pass the context here to catch cancelation event to not leak this go routine and in turn holding locks?

Reply:
@grobinson-grafana Thinking about something like this:

https://github.com/grafana/loki/compare/ingest-limits-stripe-locks...stripe-lock-ctx-cancelation